crates/cansat-lora/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+#![no_std]
+
+use embedded_hal::{self, delay, nb, serial};
+
+pub enum Error<SerialError>
+where
+    SerialError: serial::Error,
+{
+    Delay,
+    ResponseError,
+    Serial(SerialError),
+}
+
+pub struct Lora<Serial> {
+    serial: Serial,
+}
+
+impl<Serial> Lora<Serial>
+where
+    Serial: serial::nb::Write<u8> + serial::nb::Read<u8>,
+{
+    pub fn new(serial: Serial) -> Self {
+        Self { serial }
+    }
+
+    fn write_all(&mut self, cmd: &[u8]) -> Result<(), Error<Serial::Error>> {
+        for &b in cmd {
+            nb::block!(self.serial.write(b)).map_err(Error::Serial)?;
+        }
+
+        Ok(())
+    }
+
+    fn read_all(&mut self, buffer: &mut [u8]) -> Result<usize, Error<Serial::Error>> {
+        let mut ptr = 0;
+
+        loop {
+            let b = nb::block!(self.serial.read()).map_err(Error::Serial)?;
+
+            buffer[ptr] = b;
+            ptr = (ptr + 1) % buffer.len();
+
+            let response_end = buffer.ends_with(b"\r\n");
+            if response_end {
+                break;
+            }
+        }
+
+        Ok(ptr)
+    }
+
+    pub fn send<D>(
+        &mut self,
+        cmd: &[u8],
+        response_buffer: &mut [u8],
+        delay: &mut D,
+    ) -> Result<usize, Error<Serial::Error>>
+    where
+        D: delay::blocking::DelayUs,
+    {
+        self.write_all(cmd)?;
+        delay.delay_ms(20).map_err(|_| Error::Delay)?;
+        let reps_len = self.read_all(response_buffer)?;
+
+        if response_buffer.starts_with(b"ERR") {
```

Comment:
Ok, so I ran the code and turns out the way we were checking for errors in OBC was wrong :))). Some examples of an error message:
```
AT+dupa
+AT: ERROR(-10)

AT+DR=dupa
+DR: ERROR(-1)
```

See [Wio-E5 AT Command Specification](https://files.seeedstudio.com/products/317990687/res/LoRa-E5%20AT%20Command%20Specification_V1.0%20.pdf) in section 2.4.

Test your code, don't just assume it works.

So, the way to go will be to parse the error, since we want to extract and return the error code. I wrote a simple parser using [nom](https://crates.io/crates/nom) parser combinators library.
```bash
cargo add -p cansat-lora --no-default-features nom
```

```rust
use nom::{
    bytes::complete::{tag, take_until},
    character::complete::i8,
    combinator::all_consuming,
    sequence::{delimited, preceded},
};

fn parse_error(input: &[u8]) -> nom::IResult<&[u8], i8> {
    let mut parser = all_consuming(preceded(
        take_until(":"),
        delimited(tag(": ERROR("), i8, tag(")\r\n")),
    ));
    parser(input)
}

#[test]
fn test_error_parser() {
    assert_eq!(parse_error(b"+DR: ERROR(-1)\r\n"), Ok((b"".as_slice(), -1)));
}
```

Reply:
One thing I'm not sure about is that we will be writing the error response into the output buffer AND returning the error with `Result`. Maybe we should read the response into an internal buffer first, and then only copy the content if it's not an error. I don't really know. Let's leave this for later.

Reply:
Ok I made up my mind. let's not parse the message in the send function. Make the `parse::response` function public. You can't really design a nice API that does both without alloc